internal/p2p/peermanager.go

```diff
@@ -815,23 +934,89 @@ func (m *PeerManager) Advertise(peerID types.NodeID, limit uint16) []NodeAddress
 		addresses = append(addresses, m.options.SelfAddress)
 	}
 
-	for _, peer := range m.store.Ranked() {
+	var numAddresses int
+	var attempts int
+
+	ranked := m.store.Ranked()
+	seenAddresses := map[NodeAddress]struct{}{}
+
+	// get the total number of possible addresses
+	for _, peer := range ranked {
 		if peer.ID == peerID {
 			continue
 		}
 
-		for nodeAddr, addressInfo := range peer.AddressInfo {
+		for addr := range peer.AddressInfo {
+			if _, ok := m.options.PrivatePeers[addr.NodeID]; !ok {
+				numAddresses++
+			}
+		}
+	}
+
+RETRY:
+	for {
+		attempts++
+		addedLastIteration := false
+	ITERATION:
+		for idx, peer := range ranked {
+			if peer.ID == peerID {
+				continue ITERATION
+			}
+
 			if len(addresses) >= int(limit) {
-				return addresses
+				break RETRY
 			}
 
-			// only add non-private NodeIDs
-			if _, ok := m.options.PrivatePeers[nodeAddr.NodeID]; !ok {
-				addresses = append(addresses, addressInfo.Address)
+		PEER:
+			for nodeAddr, addressInfo := range peer.AddressInfo {
+				if len(addresses) >= int(limit) {
+					break RETRY
+				}
+
+				if _, ok := seenAddresses[addressInfo.Address]; ok {
+					continue PEER
+				}
+
+				// only add non-private NodeIDs
+				if _, ok := m.options.PrivatePeers[nodeAddr.NodeID]; !ok {
+					// add the peer if the total
+					// number of ranked peers is
+					// will fit within the limit,
+					// and then with decreasing
+					// liklihood if the number the
+					// lower the priority of the
+					// peers is.
+
+					// nolint:gosec // G404: Use of weak random number generator
+					if len(ranked) <= int(limit) || rand.Intn(idx+1*2) <= idx {
+						seenAddresses[addressInfo.Address] = struct{}{}
+						addresses = append(addresses, addressInfo.Address)
+						addedLastIteration = true
+					}
+
+				} else {
+					// skip private peers in future iterations
+					seenAddresses[addressInfo.Address] = struct{}{}
+				}
+
 			}
 		}
+
+		// once we have the required number of peers, we
+		// should stop.
+		if len(addresses) >= int(limit) {
+			break RETRY
+		}
+		// give up eventually
+		if attempts >= 32 && !addedLastIteration {
+			break RETRY
+		}
```

Comment:
I think in practice there aren't peers that have more than one address, because each peer can only advertise one endpoint, so the multi-address endpoint stuff is all aspirational scaffolding.

While I think your approach is probably totally fine:
- all peers except persistent peers are ranked 0 when the node starts up, because score isn't persistent
- persistent peers have very high scores
- the ranked order of peers in your implementation isn't important, and we should probably just look at all peers?

Reply:
> all peers except persistent peers are ranked 0 when the node starts up, because score isn't persistent

That seems fine, we can still generate 0 as a random value and will become more likely to do that the more higher scored peers are removed.

> the ranked order of peers in your implementation isn't important, and we should probably just look at all peers?

This is definitely true so it doesn't really matter if we use ranked or just all.

Reply:
> the lower you are in the ranked list, the less likely you are to show up in the PEX packet

fwiw, this is true of both approaches.

> I guess my approach doesn't/didn't decay as much as yours, but I've tried to clean this up.

The thing that I think is mildly preferred about the score-based approach is that you're guaranteed to get a full list after `limit` iterations. I think we can alter this approach to make it a bit closer by at least looping through list and removing the private peers before the main loop since they are irrelevant and should never be selected.

Reply:
ok, I added a fix for this, I hope.

Reply:
:eyes: